### PR TITLE
[registration] Update line search trial value selection in NDT for multi-threading

### PR DESCRIPTION
**Pull request by koide3**
Since the execution order of a multi-threaded loop can be changed, ```computeDerivatives()``` may return slightly different scores for the same input due to the floating point summation error. This can cause a condition where ```a_l == a_t``` but ```f_t != f_l``` (```f_*``` is scores, ```a_*``` are line search values) in ```trialValueSelectionMT()``` that results in yielding NaN at: https://github.com/PointCloudLibrary/pcl/blob/ff9adcf77109a1d510b51c393e8f56306123615a/registration/include/pcl/registration/impl/ndt.hpp#L481

This PR adds some condition checks in ```trialValueSelectionMT()``` to avoid this problem.

Reference: https://github.com/PointCloudLibrary/pcl/pull/4277

### CODE REVIEW

**Review by kunaltyagi — APPROVED** [review]
LGTM.

PS: Should we restructure the code since this condition is repeated 3 times? Separating the code from condition and making a jump table also makes it easier for the compiler to optimize this

```cpp
int condition = 0;  // ideally an enum with some names, but I didn't read the paper
if (a_t == a_l) { condition = 0; }
else if (f_t > f_l) { condition = 1; }
else if (g_t * g_l < 0) { condition = 2; }
else if (std::fabs (g_t) <= std::fabs (g_l)) { condition = 3; }

switch(condition) {
case 1: return thing;
case 2: return onething;
case 3: return smthg;
default:
case 0: return default_thing;
}
```

**Comment by koide3** [comment]
I updated the code. ```EndpointsCondition``` corresponds to ```Case 1 ~ Case 4``` in the paper.

File: registration/include/pcl/registration/impl/ndt.hpp [hunk 1]
```diff
@@ -473,83 +473,121 @@ NormalDistributionsTransform<PointSource, PointTarget>::trialValueSelectionMT (d
...
+    condition = EndpointsCondition::Case3;
+  }
+  else
+  {
+    condition = EndpointsCondition::Case4;
+  }
 
-    double a_t_next;
+  switch(condition) {
+    // Case 1 in Trial Value Selection [More, Thuente 1994]
```
[the remark this answers is not in the document]
Comment (koide3) [reply registration/include/pcl/registration/impl/ndt.hpp]: I removed these comments

**Review by kunaltyagi — APPROVED** [review]

**Comment by kunaltyagi** [comment]
Windows CI is failing at install C++ deps, not related to the PR

**Review by mvieth — APPROVED** [review]
Looks good, just one thought: is it sufficient to check for exact equality (`==`), or is it necessary to check for near equality, e.g. to check whether the difference of the two values is below some threshold?

**Comment by koide3** [comment]
This update aims to handle an error caused by multi-threading that happens when ```a_l == a_t```, and exact equality check is sufficient in this case.

**Pull request merged**